### PR TITLE
deprecate: optional root `project_id` attribute in `mongdbatlas_project_api_key` is deprecated

### PR DESCRIPTION
**Pull request by AgustinBettati**
## Description

As a follow up to https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1664, this PR marks the adjusted attribute as deprecated. Created in separate PR to have explicit entry in change log.

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.

### CODE REVIEW

File: mongodbatlas/resource_mongodbatlas_project_api_key.go [hunk 1]
```diff
@@ -24,8 +24,9 @@ func resourceMongoDBAtlasProjectAPIKey() *schema.Resource {
 		},
 		Schema: map[string]*schema.Schema{
 			"project_id": {
-				Type:     schema.TypeString,
-				Optional: true,
+				Type:       schema.TypeString,
+				Optional:   true,
+				Deprecated: fmt.Sprintf(DeprecationByVersionMessageParameter, "1.16.0"),
```
Comment (AgustinBettati) [inline mongodbatlas/resource_mongodbatlas_project_api_key.go]: @Zuhairahmed As this deprecation is not related to Atlas API or has any fixed date, similar to how we handled https://jira.mongodb.org/browse/INTMDB-1211 I set 3 major versions before removal. Let me know what you think.

Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_project_api_key.go]: Do we have to update the data source as well?

File: website/docs/r/project_api_key.html.markdown [hunk 2]
```diff
@@ -46,7 +46,7 @@ resource "mongodbatlas_project_api_key" "test" {
 ## Argument Reference
 
 * `description` - (Required) Description of this Project API key.
-* `project_id` - Unique 24-hexadecimal digit string that identifies your project.
+* `project_id` - Unique 24-hexadecimal digit string that identifies your project. **WARNING:** this parameter is deprecated as it no longer needs to be defined. It will be removed in version 1.16.0.
```
Comment (andreaangiolillo) [inline website/docs/r/project_api_key.html.markdown:49]: Isn't this a breaking change? 🤔

File: mongodbatlas/resource_mongodbatlas_project_api_key.go (hunk 1 above)
Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: We should update our acceptance tests to not use the deprecated field

Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_project_api_key.go]: [if is not a breaking change] Can we add a migration test scenario to make sure that upgrading to the latest version with the deprecated field does not bring an error in the plan?

File: website/docs/r/project_api_key.html.markdown (hunk 2 above)
Comment (andreaangiolillo) [reply website/docs/r/project_api_key.html.markdown:49]: if the customers have to update their configuration to not use this field -> it is a breaking change

Comment (AgustinBettati) [reply website/docs/r/project_api_key.html.markdown:49]: > if the customers have to update their configuration

Customers will not have to make any changes to there existing configuration with this change, the `project_id` is made optional and deprecated as it will be eventually deleted (this is when it will be a breaking change). For reference, HashiCorp versioning guidelines allow deprecating attributes in a minor version as existing functionality remains compatible: https://developer.hashicorp.com/terraform/plugin/best-practices/versioning#example-minor-number-increments

File: mongodbatlas/resource_mongodbatlas_project_api_key.go (hunk 1 above)
Comment (AgustinBettati) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: yes we can add this case

Comment (marcosuma) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: Can you create a jira issue for this and target 1.16.0? So we don't forget it

**Review by marcosuma — APPROVED** [review]

File: mongodbatlas/resource_mongodbatlas_project_api_key.go (hunk 1 above)
Comment (AgustinBettati) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: > Do we have to update the data source as well?

Data source implementation remains the same. This is because the minimum attributes required to obtain information of the api key at the moment is using an api_key_id together with a project_id:
```
data "mongodbatlas_project_api_key" "test" {
  api_key_id = vars.api_key_id
  project_id = vars.project_id
}
```
(this is fetching the information of the api key, with all its project assignments)

> We should update our acceptance tests to not use the deprecated field

Yes can adjust so that eventually when field is removed less changes are required. Will however leave one test with old definition to ensure no breaking changes.

Comment (AgustinBettati) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: included in https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1664

File: mongodbatlas/resource_project_api_key.go [hunk 3]
```diff
@@ -30,8 +30,9 @@ func ResourceProjectAPIKey() *schema.Resource {
 		},
 		Schema: map[string]*schema.Schema{
 			"project_id": {
-				Type:     schema.TypeString,
-				Optional: true,
+				Type:       schema.TypeString,
+				Optional:   true,
+				Deprecated: fmt.Sprintf(DeprecationByVersionMessageParameter, "1.16.0"),
```
Comment (lantoli) [inline mongodbatlas/resource_project_api_key.go]: take a look at the internal/common/constant package

Comment (AgustinBettati) [reply mongodbatlas/resource_project_api_key.go]: thanks!

**Review by lantoli — APPROVED** [review]

**Comment by AgustinBettati** [comment]
Verified failing test are not related (alert config and database user)
